hydra-explorer/src/Hydra/Explorer/ExplorerState.hs

```diff
@@ -0,0 +1,197 @@
+    Just _headState -> replaceHeadState newHeadState explorerState
+    Nothing -> explorerState <> [newHeadState]
+ where
+  newHeadState =
+    HeadState
+      { headId
+      , seedTxIn
+      , status = Initializing
+      , contestationPeriod
+      , members =
+          fmap
+            ( \(party, onChainId) ->
+                HeadMember
+                  { party
+                  , onChainId
+                  , commits = []
+                  }
+            )
+            (parties `zip` participants)
+      }
+
+aggregateAbortObservation :: AbortObservation -> ExplorerState -> ExplorerState
+aggregateAbortObservation AbortObservation{headId} explorerState =
+  case findHeadState headId explorerState of
+    Just headState ->
+      let newHeadState = headState{status = Aborted}
+       in replaceHeadState newHeadState explorerState
+    -- REVIEW: this should never happen; how should we deal with this scenario?
+    Nothing -> explorerState
+
+aggregateCommitObservation :: CommitObservation -> ExplorerState -> ExplorerState
+aggregateCommitObservation CommitObservation{headId, commitOutput, party} explorerState =
+  case findHeadState headId explorerState of
+    Just headState ->
+      let newHeadState = updateMember headState
+       in replaceHeadState newHeadState explorerState
+    -- REVIEW: this should never happen; how should we deal with this scenario?
+    Nothing -> explorerState
+ where
+  updateMember headState@HeadState{members} =
+    case find (\HeadMember{party = partyMember} -> partyMember == party) members of
+      -- REVIEW: this should never happen; how should we deal with this scenario?
+      Nothing -> headState
+      Just headMember@HeadMember{commits = currentCommits} ->
+        let (txIn, txOut) = commitOutput
+            newPartyCommit = PartyCommit{txIn, txOut}
+            newMember = headMember{commits = newPartyCommit : currentCommits}
+            newMembers = replaceMember members newMember
+         in headState{members = newMembers}
+
+  replaceMember members newMember@HeadMember{party = newHeadMember} =
+    case members of
+      -- REVIEW: this should never happen; how should we deal with this scenario?
+      [] -> [newMember]
+      (headMember@HeadMember{party = currentHeadMember} : tailMembers) ->
+        if newHeadMember == currentHeadMember
+          then newMember : tailMembers
+          else headMember : replaceMember tailMembers newMember
+
+aggregateCollectComObservation :: CollectComObservation -> ExplorerState -> ExplorerState
+aggregateCollectComObservation CollectComObservation{headId} explorerState =
+  case findHeadState headId explorerState of
+    Just headState ->
+      let newHeadState = headState{status = Open}
+       in replaceHeadState newHeadState explorerState
+    -- REVIEW: this should never happen; how should we deal with this scenario?
+    Nothing -> explorerState
+
+aggregateCloseObservation :: CloseObservation -> ExplorerState -> ExplorerState
```

Comment:
all these aggregate functions are similar, perhaps we can write more general function for aggregation?